core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -284,6 +286,11 @@ public DeleteWriteBuilder equalityFieldIds(int... fieldIds) {
       return this;
     }
 
+    public DeleteWriteBuilder withSortOrder(SortOrder newSortOrder) {
```

Comment:
If we decide to do that, we should probably cover ORC.

Reply:
Thank you for the review! I think setting an order for regular data files is covered by changes in `DataWriter` file, as in Avro/Parquet/ORC files what `WriteBuilder` builds is actually an `FileAppender` that doesn't track file properties, and it's the `DataWriter` which wraps around this file appender that actually populates these information into a `DataFile` when it closes. I do think this seems a bit confusing and we may want to refactor these classes a bit to make delete and data writers to be constructed in a more similar way.

Reply:
Oh, yeah, you are right. I think we should refactor this to make it clear (independently of this PR).

There are a few points to discuss.

The first one is `FileAppenderFactory`. It was originally created to produce `FileAppender` instances only, which made sense. Now, it also creates data and delete writers too, which makes less sense to me. I'd consider deprecating  methods for creating writers in `FileAppenderFactory` and would create `WriterFactory` instead.

```
WriterFactory {
  newDataWriter(...)
  newEqualityDeleteWriter(...)
  newPositionDeleteWriter(...) 
}
```

Second, I'd consider adding a method called `buildWriter` to our Parquet, Avro, ORC classes that would create and initialize `DataWriter`, making it consistent with our delete writers. Otherwise, it looks a bit weird.

Reply:
Thoughts, @yyanyy @rdblue @openinx @jackye1995?

Reply:
W.r.t ORC, I think we just did not have time to implement it. I'd be happy to review if there is a PR.

Reply:
I think the first sounds good to me! I think both `DataWriter` and `*DeleteWriter` are wrapper around appender, but today to construct the latter we have to specify everything in appender factories for each file format, and `Parquet` `Avro` actually directly pass them over to the underlying appender. I wonder if we can have separate `newAppender` and `newDeleteAppender` in `FileAppenderFactory` (or even generalize into one) so that we don't do this delegation per file format, and move the three methods you mentioned to `WriterFactory` to separate the two writer factories. In this case `WriterFactory` may always rely on a `FileAppenderFactory` in constructor, which might still be fine. 

For adding `buildWriter` to Parquet, Avro, ORC classes, I'm not entirely sure though, since `DataWriter` does not differ much among three file formats, so adding them to each file format may actually increase duplicated code. I haven't looked into the code deeply but I wonder if we can abstract the creation of appender out from delete builders, so that the new `WriterFactory` can create delete writers as simple as create data writers today, so that we will just have this logic in each `WriterFactory`.

Reply:
I'd need to take a closer look too. However, I do like the idea of abstracting appenders from writers. Let's discuss this in a separate issue, @yyanyy.